### PR TITLE
[All Hosts] (browsers) Information about supporting/testing add-ins with IE 11

### PR DESCRIPTION
**Pull request by Rick-Kirkham**
Provides some detail about supporting/testing with IE 11. Also, adds some more reminders that devs should test because AppSource will validate.

### CODE REVIEW

File: docs/develop/support-ie-11.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+If you plan to market your add-in through AppSource or you plan to support certain older versions of Windows and Office, your add-in must work in the embeddable browser control that is based on Internet Explorer 11 (called "Trident"). For information about which combinations of Windows and Office use the Internet Explorer 11 browser control, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
+
+> [!IMPORTANT]
+> Internet Explorer 11 doesn't support some HTML5 features such as media, recording, and location. If your add-in must support Internet Explorer 11 (see the second paragraph), then you can't use these features.
+
+Internet Explorer 11 doesn't support JavaScript versions later than ES5. If you want to use the syntax and features of ECMAScript 2015 or later, or TypeScript, you have two options as described in this article. (You can also combine these two techniques.)
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:17]: Would it be worthwhile to combine this paragraph and "Important" note? They both are describing stuff that's not supported in IE11. Maybe there could be a bulleted list of unsupported features, like:
- JavaScript versions ES6 and later
- TypeScript
- Media
- Recording
- Location

File: docs/develop/support-ie-11.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+## Use a transpiler
+
+You can write your code in either TypeScript or in a version of JavaScript later than ES5 and then transpile it at build-time into ES5 JavaScript. The resulting ES5 files are what you upload to your add-in's web application.
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:21]: Suggestion for brevity.
```suggestion
You can write your code in either TypeScript or modern JavaScript and then transpile it at build-time into ES5 JavaScript. The resulting ES5 files are what you upload to your add-in's web application.
```

File: docs/develop/support-ie-11.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+There are two popular transpilers. Both of them can work with source files that are TypeScript or post-ES5 JavaScript. They will both also work with React files (.jsx and .tsx).
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:23]:
```suggestion
There are two popular transpilers. Both of them can work with source files that are TypeScript or post-ES5 JavaScript. They also work with React files (.jsx and .tsx).
```

File: docs/develop/support-ie-11.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+- [babel](https://babeljs.io/)
+- [tsc](https://www.typescriptlang.org/index.html)
+
+See the documentation for either of them for information about installing and configuring the transpiler in your add-in project. We recommend that you use a task runner, such as [Grunt](https://gruntjs.com/) or [WebPack](https://webpack.js.org/) to automate the transpilation. For examples of add-ins that use transpilers, open [Office Add-ins Patterns and Practices](https://github.com/OfficeDev/PnP-OfficeAddins) repo and search for “target: es5” to find samples that use tsc and search for "babel" to find samples that use babel.
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:28]: Would it worthwhile just picking a sample that uses those, rather than asking people to search?

File: docs/develop/support-ie-11.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+> [!NOTE]
+> If you are using Visual Studio (not Visual Studio Code), tsc is probably easiest to use. You can install support for it with a nuget package. For more information, see [JavaScript and TypeScript in Visual Studio 2019](/visualstudio/javascript/javascript-in-vs-2019). To use babel with Visual Studio, you could create a build script and/or use the Task Runner Explorer in Visual Studio and tools like the [WebPack Task Runner](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.WebPackTaskRunner) or [NPM Task Runner](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.NPMTaskRunner).
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:31]: The style guide asks to avoid "and/or". I also find the last part of the last sentence to be confusing because of the "and", so I'm wondering if "with" makes a clearer association of the TaskRunner tools with the explorer.
```suggestion
> If you are using Visual Studio (not Visual Studio Code), tsc is probably easiest to use. You can install support for it with a nuget package. For more information, see [JavaScript and TypeScript in Visual Studio 2019](/visualstudio/javascript/javascript-in-vs-2019). To use babel with Visual Studio, create a build script or use the Task Runner Explorer in Visual Studio with tools like the [WebPack Task Runner](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.WebPackTaskRunner) or [NPM Task Runner](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.NPMTaskRunner).
```

File: docs/develop/support-ie-11.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+## Use a polyfill
+
+A [polyfill](https://en.wikipedia.org/wiki/Polyfill_(programming)) is JavaScript that duplicates some functionality that was introduced in a relatively recent version of JavaScript, but the polyfill is written in an earlier version of JavaScript; so it can run in browsers that don't support the later version. For example, the string method `startsWith` wasn't part of the ES5 version of JavaScript, and so it won't run in Internet Explorer 11; but there are polyfill libraries, written in ES5, that define and implement a `startsWith` method. We recommend the [core-js](https://github.com/zloirock/core-js) polyfill library.
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:35]: These sentences are on the long side. Consider breaking them up a bit.

```suggestion
A [polyfill](https://en.wikipedia.org/wiki/Polyfill_(programming)) is earlier-version JavaScript that duplicates functionality from more recent versions of JavaScript. The polyfill works with in browsers that don't support the later JavaScript versions. For example, the string method `startsWith` wasn't part of the ES5 version of JavaScript, and so it won't run in Internet Explorer 11. There are polyfill libraries, written in ES5, that define and implement a `startsWith` method. We recommend the [core-js](https://github.com/zloirock/core-js) polyfill library.
```

File: docs/develop/support-ie-11.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+To use a polyfill library you simply load it like any other JavaScript file or module. For example, you can use a `<script>` tag in the add-in's home page HTML file (for example `<script src="/js/core-js.js"></script>`), or you can use an `import` statement in a JavaScript file (for example, `import 'core-js';`). When the JavaScript engine sees a method like `startsWith`, it will first look to see if there is a method of that name built into the language. If there is, it will call the native method. If, and only if, the method isn't built-in, the engine will look in all loaded files for it. So, the polyfilled version isn't used in browsers that support the native version.
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:37]:
```suggestion
To use a polyfill library, load it like any other JavaScript file or module. For example, you can use a `<script>` tag in the add-in's home page HTML file (for example `<script src="/js/core-js.js"></script>`), or you can use an `import` statement in a JavaScript file (for example, `import 'core-js';`). When the JavaScript engine sees a method like `startsWith`, it will first look to see if there is a method of that name built into the language. If there is, it will call the native method. If, and only if, the method isn't built-in, the engine will look in all loaded files for it. So, the polyfilled version isn't used in browsers that support the native version.
```

File: docs/develop/support-ie-11.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+Importing the entire core-js library will import all core-js features. You can also import only the polyfills that your Office Add-in requires. For instructions about how to do this, see [CommonJS APIs](https://github.com/zloirock/core-js#commonjs-api). The core-js library has most of the polyfills that you need. There are a few exceptions detailed in the [Missing Polyfills](https://github.com/zloirock/core-js#missing-polyfills) section of the core-js documentation. For example, it doesn't support `fetch`, but you can use the [fetch](https://github.com/github/fetch) polyfill.
+
+For examples of add-ins that use a polyfill, open [Office Add-ins Patterns and Practices](https://github.com/OfficeDev/PnP-OfficeAddins) repo and search for "core.js".
```
Comment (AlexJerabek) [inline docs/develop/support-ie-11.md:41]: Same comment as earlier about linking to a specific sample.

File: docs/testing/ie-11-testing.md [hunk 9]
```diff
@@ -1,44 +1,50 @@
...
+description: 'Test your Office Add-in on Internet Explorer 11.'
+ms.date: 05/19/2021
 localization_priority: Normal
 ---
 
-# Test your Office Add-in using Internet Explorer 11
+# Test your Office Add-in on Internet Explorer 11
 
-Depending on the specifications of your add-in, you may plan to support older versions of Windows and Office, which require testing on Internet Explorer 11. This is often necessary as part of submitting your add-in to AppSource. You can use the following command line tooling to switch from more modern runtimes used by add-ins to the Internet Explorer 11 runtime for this testing.
+If you plan to market your add-in in AppSource or you plan to support certain older versions of Windows and Office, you must test the add-in using the embeddable web view control that is based on Internet Explorer 11 (called "Trident"). You can use a command line to switch from more modern runtimes used by add-ins to the Internet Explorer 11 runtime for this testing. For information about which versions of Windows and Office use the Internet Explorer 11 web view control, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
```
Comment (AlexJerabek) [inline docs/testing/ie-11-testing.md:10]: Here's the same text I suggested in an earlier comment to avoid "Trident":
```suggestion
If you plan to market your add-in through AppSource or you plan to support older versions of Windows and Office, your add-in must work in the embeddable browser control that is based on Internet Explorer 11 (IE11). You can use a command line to switch from more modern runtimes used by add-ins to the Internet Explorer 11 runtime for this testing. For information about which versions of Windows and Office use the Internet Explorer 11 web view control, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
```

File: docs/testing/ie-11-testing.md [hunk 10]
```diff
@@ -1,44 +1,50 @@
...
 
-## Pre-requisites
+> [!IMPORTANT]
+> Internet Explorer 11 does not support JavaScript versions later than ES5. If you want to use the syntax and features of ECMAScript 2015 or later, you have two options:
+>
+> - Write your code in ECMAScript 2015 (also called ES6) or later JavaScript, or in TypeScript, and then compile your code to ES5 JavaScript using a compiler such as [babel](https://babeljs.io/) or [tsc](https://www.typescriptlang.org/index.html).
+> - Write in ECMAScript 2015 or later JavaScript, but also load a [polyfill](https://en.wikipedia.org/wiki/Polyfill_(programming)) library such as [core-js](https://github.com/zloirock/core-js) that enables IE to run your code.
```
Comment (AlexJerabek) [inline docs/testing/ie-11-testing.md:16]: I like this bulletpoint summary. Consider pulling it into the other article about supporting IE11.

File: docs/testing/ie-11-testing.md [hunk 11]
```diff
@@ -1,44 +1,50 @@
...
+>
+> For more information about these options, see [Support Internet Explorer 11](../develop/support-ie-11.md).
+>
+> Also, Internet Explorer 11 does not support some HTML5 features such as media, recording, and location.
+
+> [!NOTE]
+> To test how your add-in runs in Office on the web in Internet Explorer, simply open Office on the web in Internet Explorer and [sideload the add-in](create-a-network-shared-folder-catalog-for-task-pane-and-content-add-ins.md).
```
Comment (AlexJerabek) [inline docs/testing/ie-11-testing.md:23]: Some suggestions for brevity and clarity about the browser testing (instead of the embedded browser control).
```suggestion
> To test your add-in on the Internet Explorer 11 browser, open Office on the web in Internet Explorer and [sideload the add-in](create-a-network-shared-folder-catalog-for-task-pane-and-content-add-ins.md).
```

File: docs/testing/ie-11-testing.md [hunk 12]
```diff
@@ -1,44 +1,50 @@
...
-
-2. In the root folder of your new project, run the following in the command line:
+    > [!NOTE]
+    > If you have an existing project and want to add this tooling without creating a new project, skip this step and move to the next step. 
 
-```command&nbsp;line
-npx office-addin-dev-settings webview manifest.xml ie
-```
-You should see a note in the command line that the web view type is now set to IE.
+1. In the root folder of your project, run the following in the command line. (This example assumes that your project's manifest file is in the root. If it isn't, specify the relative path to the manifest file.) You should see a message in the command line that the web view type is now set to IE.
```
Comment (AlexJerabek) [inline docs/testing/ie-11-testing.md:38]: I don't think you need the parentheses here.

```suggestion
1. In the root folder of your project, run the following in the command line. This example assumes that your project's manifest file is in the root. If it isn't, specify the relative path to the manifest file. You should see a message in the command line that the web view type is now set to IE.
```

File: docs/testing/ie-11-testing.md [hunk 13]
```diff
@@ -1,44 +1,50 @@
...
 
-> ![TIP]
-> It isn't necessary to use this tooling, but it should help debug the majority of issues related to the Internet Explorer 11 runtime. For complete robustness, you should test using a computer with a copy of Windows 7 and Office 2013 installed.
+    ```command&nbsp;line
+    npx office-addin-dev-settings webview manifest.xml ie
+    ```
 
-## Command settings
+> [!TIP]
+> It isn't necessary to use this command, but it should help debug the majority of issues related to the Internet Explorer 11 runtime. For complete robustness, you should test using computers with various combinations of Windows 7, 8, and 10 and various versions of Office. For more information, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
```
Comment (AlexJerabek) [inline docs/testing/ie-11-testing.md:45]: Should we link to that article about getting older versions of Office from here?

File: docs/testing/test-debug-office-add-ins.md [hunk 14]
```diff
@@ -1,13 +1,22 @@
...
 
-This section contains guidance about testing, debugging, and troubleshooting issues with Office Add-ins.
+This article contains guidance about testing, debugging, and troubleshooting issues with Office Add-ins.
+
+## Test cross-platform and for multiple versions of Office
+
+Office Add-ins run across major platforms, so you need to test an add-in in all the platforms where your users might be running Office. This usually includes Office on the web, Office on Windows (both subscription and one-time purchase), Office on Mac, Office on iOS, and (for Outlook add-ins) Office on Android. However, there may be some situations in which you can be sure that none of your users will be working on some platforms. For example, if you are making an add-in for a company that requires its users to work with Windows computers and subscription Office, then you don't need to test for Office on Mac or one-time purchase Windows. Note, also, that on Windows computers, the version of Windows and Office will determine which browser control is used by add-ins. For more information, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
```
Comment (AlexJerabek) [inline docs/testing/test-debug-office-add-ins.md:14]: Consider pulling the browser control note into an actual NOTE block.
```suggestion
Office Add-ins run across major platforms, so you need to test an add-in in all the platforms where your users might be running Office. This usually includes Office on the web, Office on Windows (both subscription and one-time purchase), Office on Mac, Office on iOS, and (for Outlook add-ins) Office on Android. However, there may be some situations in which you can be sure that none of your users will be working on some platforms. For example, if you are making an add-in for a company that requires its users to work with Windows computers and subscription Office, then you don't need to test for Office on Mac or one-time purchase Windows. 

> [!NOTE]
> On Windows computers, the version of Windows and Office will determine which browser control is used by add-ins. For more information, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
```

File: docs/testing/test-debug-office-add-ins.md [hunk 15]
```diff
@@ -1,13 +1,22 @@
...
+
+> [!IMPORTANT]
+> Add-ins marketed through AppSource go through a validation process that includes testing on all platforms. In addition, add-ins are tested for Office on the web with all major modern browsers, including Microsoft Edge (Chromium-based WebView2), Chrome, and Safari. Accordingly, you should test on these platforms and browsers before you submit to AppSource. For more information about validation, see [Commercial marketplace certification policies](/legal/marketplace/certification-policies), especially [section 1120.3](/legal/marketplace/certification-policies#11203-functionality), and the [Office Add-in application and availability page](../overview/office-add-in-availability.md). 
+>
+> AppSource does not use Internet Explorer or Microsoft Edge (WebView1) to test add-ins in Office on the web. But if a significant number of your users will use these two browsers to open Office on the web, then you should test with them. For more information, see [Support Internet Explorer 11](../develop/support-ie-11.md) and [Troubleshooting Microsoft Edge issues](../concepts/browsers-used-by-office-web-add-ins.md#troubleshooting-microsoft-edge-issues).
```
Comment (AlexJerabek) [inline docs/testing/test-debug-office-add-ins.md]: Let's make a clear distinction between legacy and modern Edge.

```suggestion
> AppSource does not use Internet Explorer or the legacy-version of Microsoft Edge (WebView1) to test add-ins in Office on the web. But if a significant number of your users will use these two browsers to open Office on the web, then you should test with them. For more information, see [Support Internet Explorer 11](../develop/support-ie-11.md) and [Troubleshooting Microsoft Edge issues](../concepts/browsers-used-by-office-web-add-ins.md#troubleshooting-microsoft-edge-issues).
```

**Review by AlexJerabek — CHANGES_REQUESTED** [review]
I think this looks good overall and helps clarify our validation expectations. There's some branding stuff (Trident) that should be changed before publishing.

File: docs/develop/support-ie-11.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
+---
+title: Support Internet Explorer 11
+description: 'Learn how to support Internet Explorer 11 and ES5 Javascript in your add-in.'
+ms.date: 05/19/2021
+localization_priority: Normal
+---
+
+# Support Internet Explorer 11
+
+Office Add-ins are web applications that are displayed inside IFrames when running on Office on the web. Office Add-ins are displayed using embedded browser controls when running in Office on Windows or Office on the Mac. The embedded browser controls are supplied by the operating system or by a browser installed on the user's computer.
```
[the remark this answers is not in the document]
Comment (Rick-Kirkham) [reply docs/develop/support-ie-11.md:10]: The article is in the "Plan" section of the ToC so I think some readers might be beginners to add-ins. This paragraph would help orient them.

File: docs/testing/ie-11-testing.md (hunk 10 above)
Comment (Rick-Kirkham) [reply docs/testing/ie-11-testing.md:16]: Well, these two bullets appear in at least 3 articles and they are always followed by a sentence referring readers to the supporting IE 11 article for more details. So, I don't think we want to repeat these bullets in the more detailed article itself.

File: docs/develop/support-ie-11.md [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+If you plan to market your add-in through AppSource or you plan to support certain older versions of Windows and Office, your add-in must work in the embeddable browser control that is based on Internet Explorer 11 (called "Trident"). For information about which combinations of Windows and Office use the Internet Explorer 11 browser control, see [Browsers used by Office Add-ins](../concepts/browsers-used-by-office-web-add-ins.md).
+
+> [!IMPORTANT]
+> Internet Explorer 11 doesn't support some HTML5 features such as media, recording, and location. If your add-in must support Internet Explorer 11 (see the second paragraph), then you can't use these features.
```
[the remark this answers is not in the document]
Comment (Rick-Kirkham) [reply docs/develop/support-ie-11.md]: done

**Review by AlexJerabek — APPROVED** [review]
Looks great! Thanks for making those changes.
